File: src/cli/status.rs
//! Status command - Show worktree status
//!
//! This is a local command - it works on the current workspace if you're in one,
//! or requires --workspace to specify which workspace to show.

use clap::Args;
use colored::Colorize;
use std::env;
use std::path::PathBuf;

use crate::core::{GitClient, WorktreeManager, WorkspaceManager};

#[derive(Args, Debug)]
pub struct StatusArgs {
    /// Workspace to show status for (defaults to current workspace if in one)
    #[arg(short, long, value_name = "NAME")]
    workspace: Option<String>,

    /// Show detailed information
    #[arg(short, long)]
    long: bool,
}

pub async fn execute(args: StatusArgs, manager: WorkspaceManager) -> anyhow::Result<()> {
    let git = GitClient::new();

    // Determine target workspace
    let (workspace_name, workspace_path) = if let Some(name) = args.workspace {
        // Use explicitly specified workspace
        let path = manager
            .global_config()
            .get_workspace_path(&name)
            .ok_or_else(|| {
                anyhow::anyhow!(
                    "Workspace '{}' not found. Create it with: wtp create {}",
                    name, name
                )
            })?;
        (name, path)
    } else {
        // Try to detect current workspace from current directory
        detect_current_workspace(&manager)?
    };

    if !workspace_path.exists() {
        anyhow::bail!(
            "Workspace '{}' directory does not exist at {}",
            workspace_name,
            workspace_path.display()
        );
    }

    if !workspace_path.join(".wtp").exists() {
        anyhow::bail!(
            "Workspace '{}' exists in config but the directory is missing or corrupted.",
            workspace_name
        );
    }

    println!(
        "Workspace: {} at {}",
        workspace_name.cyan().bold(),
        workspace_path.display().to_string().dimmed()
    );
    println!();

    // Load worktrees
    let worktree_manager = WorktreeManager::load(&workspace_path)?;
    let worktrees = worktree_manager.list_worktrees();

    if worktrees.is_empty() {
        println!("{}", "No worktrees in this workspace.".dimmed());
        println!();
        println!("Import a worktree with:");
        println!(
            "  {}",
            "wtp import <repo_path>".cyan()
        );
        println!();
        println!("Or switch the current repo to this workspace:");
        println!("  {}", format!("wtp switch {}", workspace_name).cyan());
        return Ok(());
    }

    if args.long {
        print_detailed_status(&git, worktrees, &workspace_path).await?;
    } else {
        print_compact_status(&git, worktrees, &workspace_path).await?;
    }

    Ok(())
}

/// Detect current workspace from current directory
/// Returns (workspace_name, workspace_path) if found
fn detect_current_workspace(
    manager: &WorkspaceManager,
) -> anyhow::Result<(String, PathBuf)> {
    let current_dir = env::current_dir()?;
    let mut check_dir = current_dir.as_path();

    loop {
        // Check if this directory has a .wtp subdirectory
        if check_dir.join(".wtp").is_dir() {
            // Find which workspace this is
            for (name, path) in manager.global_config().scan_workspaces().iter() {
                if path == check_dir {
                    return Ok((name.clone(), path.clone()));
                }
            }
            // Directory has .wtp but not registered - might be an orphan
            // Return with the directory name as workspace name
            let name = check_dir
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("workspace")
                .to_string();
            return Ok((name, check_dir.to_path_buf()));
        }

        // Move up
        match check_dir.parent() {
            Some(parent) => check_dir = parent,
            None => break,
        }
    }

    // Not in any workspace
    anyhow::bail!(
        "Not in a workspace. Either:\n\
         1. Run this command from within a workspace directory, or\n\
         2. Use --workspace <NAME> to specify the workspace"
    )
}

async fn print_compact_status(
    git: &GitClient,
    worktrees: &[crate::core::WorktreeEntry],
    workspace_path: &std::path::Path,
) -> anyhow::Result<()> {
    println!(
        "{:<30} {:<20} {}",
        "REPOSITORY".bold(),
        "BRANCH".bold(),
        "STATUS".bold()
    );

    for wt in worktrees {
        let wt_full_path = workspace_path.join(&wt.worktree_path);
        let repo_display = wt.repo.display();

        if !wt_full_path.exists() {
            println!(
                "{:<30} {:<20} {}",
                repo_display,
                wt.branch.cyan(),
                "missing".red().bold()
            );
            continue;
        }

        let status_str = match git.get_status(&wt_full_path) {
            Ok(s) => s.format_compact(),
            Err(_) => "?".to_string(),
        };

        println!(
            "{:<30} {:<20} {}",
            if repo_display.len() > 30 {
                format!("{}...", &repo_display[..27])
            } else {
                repo_display
            },
            wt.branch.cyan(),
            status_str
        );
    }

    Ok(())
}

async fn print_detailed_status(
    git: &GitClient,
    worktrees: &[crate::core::WorktreeEntry],
    workspace_path: &std::path::Path,
) -> anyhow::Result<()> {
    let separator = "\u{2500}".repeat(60);

    for wt in worktrees.iter() {
        let wt_full_path = workspace_path.join(&wt.worktree_path);
        let repo_display = wt.repo.display();

        println!("{}", separator.dimmed());
        println!("  {}", repo_display.cyan().bold());
        println!("{}", separator.dimmed());

        if !wt_full_path.exists() {
            println!(
                "  {:<10} {}",
                "Status:".bold(),
                "MISSING".red().bold()
            );
            println!();
            continue;
        }

        // Branch
        println!(
            "  {:<10} {}",
            "Branch:".bold(),
            wt.branch.cyan()
        );

        // HEAD: hash + subject + relative time
        let head_short = git.get_head_commit(&wt_full_path).unwrap_or_default();
        let subject = git
            .get_last_commit_subject(&wt_full_path)
            .unwrap_or_default();
        let rel_time = git
            .get_last_commit_relative_time(&wt_full_path)
            .unwrap_or_default();

        if !head_short.is_empty() {
            println!(
                "  {:<10} {} {} {}",
                "HEAD:".bold(),
                head_short.yellow(),
                subject,
                format!("({})", rel_time).dimmed()
            );
        }

        // Status
        match git.get_status(&wt_full_path) {
            Ok(status) => {
                println!(
                    "  {:<10} {}",
                    "Status:".bold(),
                    status.format_detail_status()
                );

                // Remote
                println!(
                    "  {:<10} {}",
                    "Remote:".bold(),
                    status.format_detail_remote()
                );
            }
            Err(e) => {
                println!(
                    "  {:<10} {}",
                    "Status:".bold(),
                    format!("error: {}", e).red()
                );
            }
        }

        // Stash
        match git.get_stash_count(&wt_full_path) {
            Ok(count) if count > 0 => {
                let entry_word = if count == 1 { "entry" } else { "entries" };
                println!(
                    "  {:<10} {}",
                    "Stash:".bold(),
                    format!("{} {}", count, entry_word).yellow()
                );
            }
            Ok(_) => {
                println!(
                    "  {:<10} {}",
                    "Stash:".bold(),
                    "none".dimmed()
                );
            }
            Err(_) => {}
        }

        println!();
    }
    println!("{}", separator.dimmed());

    Ok(())
}

File: go/cmd/status.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/eddix/wtp/internal/core"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

// NewStatusCmd returns the "wtp status" command.
func NewStatusCmd() *cobra.Command {
	var workspace string
	var long bool

	cmd := &cobra.Command{
		Use:     "status",
		Short:   "Show worktree status for a workspace",
		GroupID: GroupWorkspace,
		Args:    cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			loadedConfig, warning, err := core.LoadConfig()
			if err != nil {
				return err
			}
			if warning != "" {
				color.Yellow("%s", warning)
			}

			manager := core.NewWorkspaceManager(loadedConfig)

			// Determine target workspace.
			var workspaceName, workspacePath string
			if workspace != "" {
				workspaceName = workspace
				path, ok := manager.GlobalConfig().GetWorkspacePath(workspace)
				if !ok {
					return fmt.Errorf(
						"workspace '%s' not found. Create it with: wtp create %s",
						workspace, workspace,
					)
				}
				workspacePath = path
			} else {
				cwd, err := os.Getwd()
				if err != nil {
					return fmt.Errorf("failed to get current directory: %w", err)
				}
				name, path, err := manager.DetectCurrentWorkspace(cwd)
				if err != nil {
					return err
				}
				workspaceName = name
				workspacePath = path
			}

			if info, err := os.Stat(workspacePath); err != nil || !info.IsDir() {
				return fmt.Errorf(
					"workspace '%s' directory does not exist at %s",
					workspaceName, workspacePath,
				)
			}

			wtpDir := filepath.Join(workspacePath, core.WtpDir)
			if info, err := os.Stat(wtpDir); err != nil || !info.IsDir() {
				return fmt.Errorf(
					"workspace '%s' exists in config but the directory is missing or corrupted",
					workspaceName,
				)
			}

			cyan := color.New(color.FgCyan, color.Bold)
			dim := color.New(color.Faint)

			fmt.Print("Workspace: ")
			cyan.Print(workspaceName)
			fmt.Print(" at ")
			dim.Println(workspacePath)
			fmt.Println()

			// Load worktrees.
			wtMgr, err := core.LoadWorktreeManager(workspacePath, nil)
			if err != nil {
				return err
			}
			worktrees := wtMgr.ListWorktrees()

			if len(worktrees) == 0 {
				dim.Println("No worktrees in this workspace.")
				fmt.Println()
				fmt.Println("Import a worktree with:")
				fmt.Print("  ")
				cyan.Println("wtp import <repo_path>")
				fmt.Println()
				fmt.Println("Or switch the current repo to this workspace:")
				fmt.Print("  ")
				cyan.Printf("wtp switch %s\n", workspaceName)
				return nil
			}

			git := core.NewGitClient()

			if long {
				return printDetailedStatus(git, worktrees, workspacePath)
			}
			return printCompactStatus(git, worktrees, workspacePath)
		},
	}

	cmd.Flags().StringVarP(&workspace, "workspace", "w", "", "Workspace to show status for")
	cmd.Flags().BoolVarP(&long, "long", "l", false, "Show detailed information")

	return cmd
}

func printCompactStatus(git *core.GitClient, worktrees []core.WorktreeEntry, workspacePath string) error {
	bold := color.New(color.Bold)
	cyan := color.New(color.FgCyan)
	red := color.New(color.FgRed, color.Bold)

	bold.Printf("%-30s %-20s %s\n", "REPOSITORY", "BRANCH", "STATUS")

	for _, wt := range worktrees {
		wtFullPath := filepath.Join(workspacePath, wt.WorktreePath)
		repoDisplay := wt.Repo.Display()

		if info, err := os.Stat(wtFullPath); err != nil || !info.IsDir() {
			// Truncate long names.
			if len(repoDisplay) > 30 {
				repoDisplay = repoDisplay[:27] + "..."
			}
			fmt.Printf("%-30s ", repoDisplay)
			cyan.Printf("%-20s ", wt.Branch)
			red.Println("missing")
			continue
		}

		statusStr := "?"
		if status, err := git.GetStatus(wtFullPath); err == nil {
			statusStr = status.FormatCompact()
		}

		// Truncate long names.
		if len(repoDisplay) > 30 {
			repoDisplay = repoDisplay[:27] + "..."
		}

		fmt.Printf("%-30s ", repoDisplay)
		cyan.Printf("%-20s ", wt.Branch)
		fmt.Println(statusStr)
	}

	return nil
}

func printDetailedStatus(git *core.GitClient, worktrees []core.WorktreeEntry, workspacePath string) error {
	bold := color.New(color.Bold)
	cyan := color.New(color.FgCyan, color.Bold)
	yellow := color.New(color.FgYellow)
	red := color.New(color.FgRed, color.Bold)
	dim := color.New(color.Faint)

	separator := "\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500" +
		"\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500" +
		"\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500" +
		"\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500" +
		"\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500" +
		"\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500"

	for _, wt := range worktrees {
		wtFullPath := filepath.Join(workspacePath, wt.WorktreePath)
		repoDisplay := wt.Repo.Display()

		dim.Println(separator)
		fmt.Print("  ")
		cyan.Println(repoDisplay)
		dim.Println(separator)

		if info, err := os.Stat(wtFullPath); err != nil || !info.IsDir() {
			fmt.Print("  ")
			bold.Print("Status:   ")
			red.Println("MISSING")
			fmt.Println()
			continue
		}

		// Branch.
		fmt.Print("  ")
		bold.Print("Branch:   ")
		color.New(color.FgCyan).Println(wt.Branch)

		// HEAD: hash + subject + relative time.
		headShort, _ := git.GetHeadCommit(wtFullPath, true)
		subject, _ := git.GetLastCommitSubject(wtFullPath)
		relTime, _ := git.GetLastCommitRelativeTime(wtFullPath)

		if headShort != "" {
			fmt.Print("  ")
			bold.Print("HEAD:     ")
			yellow.Print(headShort)
			fmt.Printf(" %s ", subject)
			dim.Printf("(%s)\n", relTime)
		}

		// Status.
		status, err := git.GetStatus(wtFullPath)
		if err != nil {
			fmt.Print("  ")
			bold.Print("Status:   ")
			red.Printf("error: %v\n", err)
		} else {
			fmt.Print("  ")
			bold.Print("Status:   ")
			fmt.Println(status.FormatDetailStatus())

			fmt.Print("  ")
			bold.Print("Remote:   ")
			fmt.Println(status.FormatDetailRemote())
		}

		// Stash.
		stashCount, err := git.GetStashCount(wtFullPath)
		if err == nil {
			fmt.Print("  ")
			bold.Print("Stash:    ")
			if stashCount > 0 {
				entryWord := "entries"
				if stashCount == 1 {
					entryWord = "entry"
				}
				yellow.Printf("%d %s\n", stashCount, entryWord)
			} else {
				dim.Println("none")
			}
		}

		fmt.Println()
	}

	dim.Println(separator)
	return nil
}

func init() {
	rootCmd.AddCommand(NewStatusCmd())
}
